demos/TicTacToe/Classes/TicTacToeScene.cpp

```diff
@@ -1,26 +1,271 @@
+                      const char* previous_sibling) override {
+    LogMessage("  ChildListener.OnChildChanged(%s)", snapshot.key());
+    events_.push_back(std::string("changed ") + snapshot.key());
+  }
+  void OnChildMoved(const firebase::database::DataSnapshot& snapshot,
+                    const char* previous_sibling) override {
+    LogMessage("  ChildListener.OnChildMoved(%s)", snapshot.key());
+    events_.push_back(std::string("moved ") + snapshot.key());
+  }
+  void OnChildRemoved(
+      const firebase::database::DataSnapshot& snapshot) override {
+    LogMessage("  ChildListener.OnChildRemoved(%s)", snapshot.key());
+    events_.push_back(std::string("removed ") + snapshot.key());
+  }
+  void OnCancelled(const firebase::database::Error& error_code,
+                   const char* error_message) override {
+    LogMessage("ERROR: SampleChildListener canceled: %d: %s", error_code,
+               error_message);
+  }
+
+  // Get the total number of Child events this listener saw.
+  size_t total_events() { return events_.size(); }
+
+  // Get the number of times this event was seen.
+  int num_events(const std::string& event) {
+    int count = 0;
+    for (int i = 0; i < events_.size(); i++) {
+      if (events_[i] == event) {
+        count++;
+      }
+    }
+    return count;
+  }
+
+ public:
+  // Vector of strings defining the events we saw, in order.
+  std::vector<std::string> events_;
+};
+
+// A ValueListener that expects a specific value to be set.
+class ExpectValueListener : public firebase::database::ValueListener {
+ public:
+  explicit ExpectValueListener(firebase::Variant wait_value)
+      : wait_value_(wait_value.AsString()), got_value_(false) {}
+  void OnValueChanged(
+      const firebase::database::DataSnapshot& snapshot) override {
+    if (snapshot.value().AsString() == wait_value_) {
+      got_value_ = true;
+    } else {
+      LogMessage(
+          "FAILURE: ExpectValueListener did not receive the expected result.");
+    }
+  }
+  void OnCancelled(const firebase::database::Error& error_code,
+                   const char* error_message) override {
+    LogMessage("ERROR: ExpectValueListener canceled: %d: %s", error_code,
+               error_message);
+  }
+
+  bool got_value() { return got_value_; }
+
+ private:
+  firebase::Variant wait_value_;
+  bool got_value_;
+};
+
+// Wait for a Future to be completed. If the Future returns an error, it will
+// be logged.
+void WaitForCompletion(const firebase::FutureBase& future, const char* name) {
+  while (future.status() == firebase::kFutureStatusPending) {
+    ProcessEvents(100);
+  }
+  if (future.status() != firebase::kFutureStatusComplete) {
+    LogMessage("ERROR: %s returned an invalid result.", name);
+  } else if (future.error() != 0) {
+    LogMessage("ERROR: %s returned error %d: %s", name, future.error(),
+               future.error_message());
+  }
+}
+char* random_string(std::size_t length) {
```

Comment:
Returning a raw pointer that you've allocated is a dangerous thing to do. It's not clear who owns this pointer, or who is responsible for cleaning it up when it's no longer being used. 

A better solution would be to either pass in a buffer and size so that it can be populated, then it's clearer who owns that buffer (the caller) or to return a std::string or std::unique_ptr (probably a std::string in this case).